design/keycloak.x/operator.md

```diff
@@ -0,0 +1,128 @@
+# Keycloak.X Operator
+
+## Motivation for a new operator
+
+The current Operator made in Go Lang served us and the community well so far, but increasing challenges are paving the road for a complete re-write.
+
+* The codebase is hard to maintain because of organic growth and accumulated technical debt
+* The Keycloak community is more keen to Java and there is less Go Lang expertise
+* The current project needs some high-cost maintenance tasks to be performed in order to use the latest features (e.g. webhooks) and receive the latest fixes and patches, specifically:
+    * upgrading Go lang version from 1.13 to 1.17
+    * upgrading the Operator SDK and the dependencies
+
+  Those upgrades will require creating a completely new project, using different libraries, moving, and in some cases, rewriting the components, e.g. the whole testsuite.
+* The current approach around CRDs no longer fits the long term vision for cloud-native Keycloak as it is very error-prone and fragile.
+* A Java operator can share business objects with the Keycloak main codebase increasing the code-reuse and dramatically reducing the chances of introducing bugs in the translation process from Kubernetes resources.
+* A unified codebase will make it easy to test and debug the entire system.
+* A new operator will embrace from the ground up the new Cloud Native capabilities of upcoming Keycloak releases such as the Quarkus distribution and Store.X, making those first-class citizen overall improving the user experience.
+
+
+## Features
+
+---
+**NOTE**
+
+The primary target of the operator is to make it easy to achieve production grade installations of Keycloak.X.
+
+---
+
+### Configuring Keycloak deployment
+
+The operator will use a CRD representing the Keycloak installation. The CRD will expose the following configuration options:
```

Comment:
I'm missing the following configuration options:

* Clustering. Considering the current storage we might want to have options to set the number of `owners` for distributed caches, the number of entries, etc. Not sure how that would look like with the new store though, especially because the operator might need an Infinispan Server in place.
* JVM Settings. Would the operator provide specific options for setting specific JVM options (heap size, new size, gc, etc) or a single option to pass any JVM setting you want? I know this is something we should try to avoid as much as possible, but IMO this is a must (e.g.: capture gc logs)

Reply:
+1 for this. Imo Caching/Clustering as well as the JVM environment is an infrastructure concern and thus should be handled by the operator. Not sure how Store.X fits in here, though.

Reply:
I believe both areas are basically a dist config options and as such are configured as any other Quarkus dist option, that means env vars e.g. Is that correct? We have this already covered by:
> Since most of the configuration options will come from the Keycloak.X distribution itself, the CRD will also expose appropriate fields for passing any distribution related options to the container